### ISSUE TITLE
RPMDIntegrator does not integrate at double precision at mixed prec mode

### ISSUE THREAD

**Issue by cihe13375** [openmm/openmm#4077]
Hi, 

According to documentation:
> Precision: This selects what numeric precision to use for calculations. ... If it is set to “mixed”, forces are computed in single precision but integration is done in double precision. This gives much better energy conservation with only a slight decrease in speed.

However, this is not true for the `RPMDIntegrator`. Namely, the total energy drops at the relative rate of ~1e-7/step at NVE ensemble (`setApplyThermostat=False`), even with no external force and `stepSize=0`. This issue makes long NVE simulations unfeasible.

After digging into the source code, I believe the problem mostly lies in the `cc.doubleToString()` function. This function is heavily used in `CommonRpmdKernels.cpp`, especially in the FFT section, However, this function returns `float` values at mixed precision mode:
```
string ComputeContext::doubleToString(double value) const {
    stringstream s;
    s.precision(getUseDoublePrecision() ? 16 : 8);
    s << scientific << value;
    if (!getUseDoublePrecision())
        s << "f";
    return s.str();
}
```
It results in the whole FFT having only the accuracy of `single` as well as the velocities and energy, even if all other computations are performed in `mixed` variable type.
A solution is to add an alternative `doubleToString` function to `ComputeContext.cpp`:
```
string ComputeContext::doubleToStringAnother(double value) const {
    stringstream s;
    s.precision( (getUseDoublePrecision() || getUseMixedPrecision()) ? 16 : 8);
    s << scientific << value;
    if ( !(getUseDoublePrecision() || getUseMixedPrecision()) )
        s << "f";
    return s.str();
}
```
and replace all  `cc.doubleToString()` in `CommonRpmdKernels.cpp` to this `cc.doubleToStringAnother()`. After that the energy of NVE does not drop anymore with no external force and `stepSize=0`, and is reasonably conserved in a real simulation.

Side note: the `copyDataFromContext` and `applyCellTranslations` functions in `kernels/rpmd.cc` are also somewhat suspicious to me, because they effectively assign `real` positions (from Context) to `mixed` positions in the Integrator. However, it seemingly causes no problems in our tests.

**Comment by peastman** [comment]
You're absolutely right.  When `doubleToString()` is called by an integrator, it should produce a double precision constant in mixed precision mode.  Let me fix that!

**Comment by peastman** [comment]
> the copyDataFromContext and applyCellTranslations functions in kernels/rpmd.cc are also somewhat suspicious to me, because they effectively assign real positions (from Context) to mixed positions in the Integrator.

A bit of clarification about this.  In mixed precision mode, the Context stores positions as a single precision value, which is what the force routines load, plus a single precision correction that only the integrators look at.  Together they give approximately double precision accuracy.  `copyDataToContext()` is called to prepare for computing forces.  Since the force routines don't look at the correction, it doesn't need to be copied.

**Comment by cihe13375** [comment]
> The fix is in #4079. Can you try it out and see if it fixes the problem for you?

Thanks for the fix, can confirm it resolves the issue.

Side note: somehow I found another issue of RPMDIntegrator in the meantime (though I think they are not directly related); see #4080.

**Comment by cihe13375** [comment]
> > the copyDataFromContext and applyCellTranslations functions in kernels/rpmd.cc are also somewhat suspicious to me, because they effectively assign real positions (from Context) to mixed positions in the Integrator.
> 
> A bit of clarification about this. In mixed precision mode, the Context stores positions as a single precision value, which is what the force routines load, plus a single precision correction that only the integrators look at. Together they give approximately double precision accuracy. `copyDataToContext()` is called to prepare for computing forces. Since the force routines don't look at the correction, it doesn't need to be copied.

I mostly refer to the `copyDataFromContext` (and similarly `applyCellTranslations`), because the code is like this:
```
KERNEL void copyDataFromContext(GLOBAL mm_long* srcForce, GLOBAL mm_long* dstForce, GLOBAL mixed4* srcVel, GLOBAL mixed4* dstVel,
        GLOBAL real4* srcPos, GLOBAL mixed4* dstPos, GLOBAL int* order, int copy) {
    const int base = copy*PADDED_NUM_ATOMS;
    for (int particle = GLOBAL_ID; particle < NUM_ATOMS; particle += GLOBAL_SIZE) {
        int index = order[particle];
        dstForce[base*3+index] = srcForce[particle];
        dstForce[base*3+index+PADDED_NUM_ATOMS] = srcForce[particle+PADDED_NUM_ATOMS];
        dstForce[base*3+index+PADDED_NUM_ATOMS*2] = srcForce[particle+PADDED_NUM_ATOMS*2];
        dstVel[base+index] = srcVel[particle];
        real4 posq = srcPos[particle];
        dstPos[base+index].x = posq.x;
        dstPos[base+index].y = posq.y;
        dstPos[base+index].z = posq.z;

    }
}
```
This is called every step in `CommonIntegrateRPMDStepKernel::computeForces`, and the source code above seemingly indicates that the `dstPos`(mixed numbers in the Integrator) is directly override by `srcPos`(real numbers from Context), so the positions are truncated. Or I missed something here?
